cmd/kubeadm/app/apis/kubeadm/types.go

```diff
@@ -149,6 +149,15 @@ type NodeConfiguration struct {
 
 	// FeatureGates enabled by the user
 	FeatureGates map[string]bool
+
+	// Master flag specifies that the node should be joined as an additional master.
+	Master bool
+
+	// MasterConfigurationFile specifies the path of the cluster configuration file.
+	// Please note that the user is requested provide this file only if the
+	// RBAC rules for accessing the kubeadm-config ConfigMap during the discovery phase
+	// are explicitly removed.
+	MasterConfigurationFile string
```

Comment:
I'm confused about this... I thought we were going to rely on grabbing the configmap when joining.

Reply:
Let me try to argument better why I implemented this:

The `kubeadm join` use bootstap-tokens in its early stages, and even if there are several mechanism for making bootstap tokens secure enough, `kubeadm join` supports also some variations at the join workflow allowing the user to "[Secure the installation even more](https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-join/#securing-more)".

One of this variations, currently allows the users to turn down the access to the `cluster-info` config map (see [here](https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-join/#turning-off-public-access-to-the-cluster-info-configmap)). In that case the user is requested to provide manually the `cluster-info` via a file using `kubeadm join --discovery-file`.

The flag `MasterConfigurationFile`, was implemented for being consistent with the above conditions: use kubeadm join without getting access to any kind of information on the cluster. 

More specifically, In case the user wants to turn down the access to the `kubeadm-config` config map for bootstrap tokens, the user will be requested to provide manually the `kubeadm-config` via a file using `kubeadm join --join --master --masterConfigurationFile`.

Let me know if we should keep this or remove it. 
If we remove this, we should document that the `kubeadm join --join --master` will be supported only if the access to `kubeadm-config` config map is granted to bootstrap-tokens.

/cc @mattmoyer WDYT?